### ISSUE TITLE
Is there a plan to implement early stopping in model training?

### ISSUE THREAD

**Issue by aizest** [RasaHQ/rasa#6693]
Do we have a plan for that, or it's already there? That will save some training time and avoid overfitting issues.

**Comment by sara-tagger** [comment]
Thanks for submitting this feature request 🚀 @melindaloubser1 will get back to you about it soon! ✨

**Comment by tttthomasssss** [comment]
Hi @aizest, thanks very much for raising this question!

Do you mean early stopping as in stop training if validation performance didn't improve by at least `delta` for `n` number of epochs?

We have merged model checkpointing into master just recently (see the relevant PR [here](https://github.com/RasaHQ/rasa/pull/5985)). If model checkpointing is enabled the best model during training will be saved to disk instead of the last one. If enabled the log will also tell you at what epoch the best model occurred, which can give you some indication if you're training for too long (or indeed not long enough).

I appreciate this is not the same as early stopping, but it yields a similar outcome.

**Comment by aizest** [comment]
Hi @tttthomasssss ,

Thanks for the quick reply.

Yes. By early stopping, I mean to stop training earlier if the performance doesn't get improved in N epochs.
Here, could we specify a separate validation set to measure the performance of the latest model? 
It's better to use a validation set rather than the training set in evaluating the evolving model.
This will ensure the model has better generalizability on unseen data.

I understand the purpose to the suggested PR, which saves the best model rather than the latest model.
I agree with the design of this new feature, just want to remind the evaluation is done on a validation set.

Thanks!

**Comment by tttthomasssss** [comment]
So, at the moment we don't support early stopping (but, alas, we now have an issue for it ;)). For model checkpointing we currently use a random sample of the training set to estimate performance - which as you say is not an ideal situation, however we are working on adding support for specifying a validation set when running e.g. `rasa train ...` (see #5747).

**Comment by koernerfelicia** [comment]
Closing this as it's a duplicate of https://github.com/RasaHQ/rasa/issues/7721

**Comment by joejuzl** [comment]
#7721 is about manually stopping the training process, whereas this is about automatically determining that the process can be stopped. It seems like a different issue to me.

**Comment by koernerfelicia** [comment]
@joejuzl good catch, yes you're right

**Comment by bradyneal** [comment]
Does Rasa still not have some way to see validation set performance over time? This seems like a pretty big issue; how will we know if the model is overfitting or not?

**Comment by koernerfelicia** [comment]
@bradyneal there are other ways to validate your model similar to what you would use a validation set during training for.

You can monitor some component's  (`DIETClassifier`, `ResponseSelector`, `TEDPolicy`) performance on a held-out set using the parameters:
`evaluate_every_number_of_epochs` and `evaluate_on_number_of_examples`. Results will be logged as the component trains. This is virtually the same as being able to supply a validation set, except that you have no control over which examples will be held out.

You can also use model checkpointing to select the best model that was trained.

You can also run `rasa test` on a held-out set once you have selected a model, to check that the model hasn't overfit.

**Comment by bradyneal** [comment]
@koernerfelicia

> You can monitor some component's (DIETClassifier, ResponseSelector, TEDPolicy) performance on a held-out set using the parameters:
evaluate_every_number_of_epochs and evaluate_on_number_of_examples. Results will be logged as the component trains. This is virtually the same as being able to supply a validation set, except that you have no control over which examples will be held out.

Brilliant! 

> You can also use model checkpointing to select the best model that was trained.

I saw discussion above in this thread that mentioned the merged PR did this selection using the **training set**. Are you saying it now works with held-out data, or it still works with training data?

**Comment by koernerfelicia** [comment]
> I saw discussion above in this thread that mentioned the merged PR did this selection using the training set. Are you saying it now works with held-out data, or it still works with training data?

This one also works with the parameters I mentioned above. It will also hold some examples from the training set out, again you have no control over exactly which examples, but can specify the proportion of examples used for training and examples used for validation
